### DISCUSSION TITLE
The new era of Rust / Go based build tools

### DISCUSSION THREAD

**Discussion by Franziskus1988** [nrwl/nx#8704] [General]
Hi,

I'm currently thinking about adopting our big Angular code base to Nx. Mainly to get help for achieving good modularization, which get's more and more important with a growing code base. Unfortunately this topic was a little bit neglected by the Web community in the past 😢 
While thinking about porting to Nx I see the Rust and Go based build tools like [esbuild](https://esbuild.github.io/) and [swc](https://swc.rs/) popping up.  I'm extremely excited about these after spending in the last years much time watching the angular-cli doing it's job 😭 

What does the Nrwl team think about this? Are there already plans to also port thinks to Go / Rust? Are there plans to adopt one of the tools?

**Comment by vadistic** [comment]
Hey, I'm just exploring nx capabilities right now. I see that there is swc support in `@nrwl/js:swc` and `@nrwl/web:rollup` + compiler option.

https://nx.dev/web/package#compiler
https://nx.dev/js/swc#nrwljsswc

So I would say Nrwl team had plans about that and it's quite already there.

**Comment by Brian-McBride** [comment]
@Franziskus1988  I really need to refactor this using the new devkit libs from Nx, but here is a stab at an esbuild plugin
[@anatine/esbuildnx](https://www.npmjs.com/package/@anatine/esbuildnx)

I was mainly using it to compile Node apps down to the minimal size for deployment to GCP Cloud Functions, AWS Lambdas, etc...  although there is a benefit of building docker containers faster.

The largest challenge with Esbuild, which is less likely on a web app, is the lack of support for decorators and it will just fail out on modules it can't treeshake/compile. So you have to create a list of npm packages to exclude ahead of time. When I made this, there was no way to handle it during build. Otherwise I'd take failing modules, list them in `.cache` and then copy them into the dist node_modules folder. Current, any packages listed to exclude will copy over and look for their deps as well.

**Comment by vadistic** [comment]
@Brian-McBride Thanks for sharing this - I also implemented my version of nx esbuild tooling - your implementation will come useful to improve it :)

I see that you also handled incremental builds - this is super useful since it works with `@nrwl/node:serve` :)

Here's my stab at similar thing. I more or less know what I do not want to bundle - I can list those packages explicitly as externals. 

The challenge was to create publishable libraries with accurate packages- I needed to automatically decide when to bundle ESM-only libs when compiling to CJS or whether to include private-only libraries. Then I can just use those packages as an input for custom esbuild  node externals plugin.

So,
part 1: generate accurate package json: https://gist.github.com/vadistic/def14c8bc5300a35c509a1959131e309
part 2: implement esbuild node-externals plugin: https://gist.github.com/vadistic/3ff219cfa19dd81b9d59355d1b1470b5

As a bonus I have automatic esm specifier resolver for externals  :)
